### PR TITLE
TASK: Fusion performance optimization (lazy Component props)

### PR DESCRIPTION
**Pull request by hlubek**
**What I did**

* Components provide a nice way to structure Fusion code but prevent lazy evaluation as Fusion does by default by eagerly evaluating all properties as `props`
* If conditions are used inside the `renderer`, it's quite probable that not all props are used - so a large amount of unnecessary evaluations could be performed (we measured that to be in the range of 20-30% in a larger project)

**How I did it**

* Introduced a `LazyProps` object implementing `ArrayAccess` that evaluates the actually accessed props lazily (and caches the results)

**How to verify it**

* Components work as before (there could be edge-cases where explicit `array` type annotations are used for `props`)
* props are only evaluated if used (verified by test)

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)

Resolves: #2793

### CODE REVIEW

**Comment by mficzel** [comment]
Thinking about this further i come to the conclusion that @apply will have to learn how to deal with a lazy proxy.

**Comment by hlubek** [comment]
@mficzel Good point, I'll add a test to cover that case and try to fix it in the Runtime

**Comment by mficzel** [comment]
Would be great to make @apply support generic objects that implement ArrayAccess. That would allow to use DomainObjects for that in addition to the lazyProxy.

**Comment by hlubek** [comment]
@mficzel I think we'd need to change the way apply works by having some kind of chain (evaluate key -> isset on apply value 1 -> isset on apply value 2 -> isset in Fusion configuration), which might make things slower again... And we'd loose the possibility to enumerate all defined keys for array based implementations unless `Traversable` is implemented.

**Comment by mficzel** [comment]
@hlubek i think by enabling the lazyEvaluationProxyToTell to know which keys exist without actually evluating them yet we can have both. This is the only thing @apply needs to decide wether it should try to read this.

**Comment by hlubek** [comment]
@mficzel This is what I did in aa168b7:

```
                    elseif ($singleApplyValues instanceof LazyProps) {
                        for ($singleApplyValues->rewind(); ($key = $singleApplyValues->key()) !== null; $singleApplyValues->next()) {
                            $combinedApplyValues[$fusionPath . '/' . $key] = [
                                'key' => $key,
                                'value' => function () use ($singleApplyValues, $key) {
                                    return $singleApplyValues[$key];
                                },
                                'lazy' => true
                            ];
                        }
                    }
```

Actually we could use `Traversable` and `ArrayAccess` as interfaces to check for this case, since these provide the necessary methods and array access to to the actual lazy evaluation. So every domain object implementing these interfaces would work (and having a dependency to `FusionObjects/Helpers` from `Core` is actually a code smell).

**Comment by mficzel** [comment]
I really like the idea to use `ArrayAccess` and `Traversable` and let `LazyProps` implement those.

File: Neos.Fusion/Classes/FusionObjects/Helpers/LazyProps.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+namespace Neos\Fusion\FusionObjects\Helpers;
+
+use Neos\Flow\Annotations as Flow;
+use Neos\Fusion\Core\Runtime;
+
+/**
+ * @Flow\Proxy(false)
+ */
+final class LazyProps implements \ArrayAccess, \Iterator
```
Comment (mficzel) [inline Neos.Fusion/Classes/FusionObjects/Helpers/LazyProps.php]: Does it make sense to add this as `LazyEvaluationProxy` to the Fusion/Core? That would allow to use it in other places aswell.

**Comment by hlubek** [comment]
@mficzel It's now implemented like this. The failing test comes from UriBuilder type hints that are not reflected in the Fusion objects (null vs. empty array). Is somebody fixing this?

File: Neos.Fusion/Classes/FusionObjects/Helpers/LazyProps.php (hunk 1 above)
Comment (hlubek) [reply Neos.Fusion/Classes/FusionObjects/Helpers/LazyProps.php]: I thought about it too, but this is rather specific for holding an array of keys and defering the evaluation of these. I'm not sure where and how we would re-use this. By only checking the interfaces in Runtime we always have the chance to extract a common class or something else later.

**Comment by mficzel** [comment]
@hlubek regarding the uri-builder there is a pr to fix it already #2733

**Comment by mficzel** [comment]
@hlubek the master is green again can you update the pr (rebase or merge master)

File: Neos.Fusion/Classes/FusionObjects/Helpers/LazyProps.php (hunk 1 above)
Comment (mficzel) [reply Neos.Fusion/Classes/FusionObjects/Helpers/LazyProps.php]: :-) i now have a usecase for that. Evaluating only the current page of a multipage form. But since i the fusion-forms shall support Neos 4.3 i will copy that code anyways but thanks for the inspiration.

**Comment by bwaidelich** [comment]
I added an issue for this in order to simplify release management and to collect higher-level details for the release notes: #2793

**Comment by hlubek** [comment]
> @hlubek sadly this has conflicts with the way Neos.SEO uses apply in the line
> 
> ```
> @apply.attributes = ${props.attributes ? Array.filter(props.attributes, attribute => attribute != null) : []}
> ```

Yes, I noticed it too in the demo site. Since `attributes` is another `props` value this one fails due to the type annotation in `Array.filter`. I'd suggest to make `ArrayHelper` methods more flexible here and accept any `Traversable` value. I didn't find any other solution so far.

What do you think?

**Comment by bwaidelich** [comment]
I didn't dig into the whole topic yet, but re:

> I'd suggest to make ArrayHelper methods more flexible here and accept any Traversable value

I'm very much in favor of that. We should keep Eel helpers more relaxed in general I think

**Comment by mficzel** [comment]
@hlubek making the array helper Traversable or iterable tolerant would be great anyways. Should be done in a separate pr that would habe to be merged first.

I suggest not to rush this into 5.1 as this also breaks other things see: https://github.com/neos/neos-development-collection/pull/2738#pullrequestreview-303330436

**Comment by hlubek** [comment]
> @hlubek making the array helper Traversable or iterable tolerant would be great anyways. Should be done in a separate pr that would habe to be merged first.
> 
> I suggest not to rush this into 5.1 as this also breaks other things see: [#2738 (review)](https://github.com/neos/neos-development-collection/pull/2738#pullrequestreview-303330436)

Well, this one should be fixed by aa168b7cabc422430b9668a8368e6a2980ffefe9, isn't it?

**Comment by mficzel** [comment]
> Well, this one should be fixed by aa168b7, isn't it?

No it is still there, tested with the latest version of the pr and the examples from https://github.com/neos/neos-development-collection/pull/2738#pullrequestreview-303330436

**Comment by mficzel** [comment]
Could probably be fixed by passing down a proxy directly if it is applied.

**Comment by hlubek** [comment]
> No it is still there, tested with the latest version of the pr and the examples from [#2738 (review)](https://github.com/neos/neos-development-collection/pull/2738#pullrequestreview-303330436)

Must be an AFX issue:
```
renderer = ${props.applyValue}
```

Works as expected, even with one level more of indirection:

```
apply.renderWithNestedProps = Neos.Fusion:Component {
  applyValue = "example"
  applyError = Neos.Fusion:NotImplemented

  renderer = Neos.Fusion:TestNestedPropsA {
    applyValue = ${props.applyValue}
    applyError = ${props.applyError}
  }
}

prototype(Neos.Fusion:TestNestedPropsA) < prototype(Neos.Fusion:Component) {
  applyValue = null

  renderer = afx`
    <Neos.Fusion:TestNestedPropsB {...props} />
  `
}

prototype(Neos.Fusion:TestNestedPropsB) < prototype(Neos.Fusion:Component) {
  applyValue = null

  renderer = ${props.applyValue}
}
```

But:

```
renderer = afx`
    :: {props.applyValue} ::
  `
```

just gives an empty alue.

**Comment by hlubek** [comment]
I found the issue, it seems the stack approach for apply values breaks here with lazy evaluation (since the original apply values on the stack are not "current" anymore when the expression is evaluated). The solution: don't use a stack. IMHO it's not needed - see 0642115d529237bd30ae1502a78409e7739c4718.

**Comment by hlubek** [comment]
See neos/flow-development-collection#1871 for a fix of `Array.filter(...)` with `LazyProps`.

**Comment by mficzel** [comment]
@hlubek I am not sure that the apply stack is really obsolete. Fusion often dives deeply into rendering subcomponents and comes back again. I am pretty sure that it makes sense to keep the already evaluated apply values at hand for this case. 

Are you sure this would not lead to unneeded evaluations of "prepareApplyValuesForFusionPath". Might be not measurable in project code yet as `@apply` is still not used that often.

**Comment by hlubek** [comment]
@mficzel That's what my change does basically. It only pops the (absolute) paths if the runtime returns from an evaluation. So what changes is that parent apply values are still present (if accessed by absolute path!).

All the unit tests pass (including the new case from you) but I didn't test it in a real installation yet.

**Comment by mficzel** [comment]
@hlubek that is exactly the type of change that may work fine for isolated tests but causes trouble in complex scenarios. The fusion test coverage is quite good but i am not sure we would catch such a problem. I would like to understand this much deeper before merging and not rush this into 5.1 so i suggest to finalize this now but merge to master after 5.1.

I will have to find timt to test this in a real world project. Then i can take a look wether the applyStack is really not needed anymore.

**Comment by hlubek** [comment]
Yes ;) It's quite complex how the `Runtime` with lazy evaluation and `@apply` works. Let's see if we can find cases where this breaks and improve the test coverage.

**Comment by hlubek** [comment]
> @hlubek sadly this has conflicts with the way Neos.SEO uses apply in the line
> 
> ```
> @apply.attributes = ${props.attributes ? Array.filter(props.attributes, attribute => attribute != null) : []}
> ```

If #1871 is merged this case should be no problem.

Would be great to test give this change in a real-world project.

**Comment by kitsunet** [comment]
> > @hlubek sadly this has conflicts with the way Neos.SEO uses apply in the line
> > ```
> > @apply.attributes = ${props.attributes ? Array.filter(props.attributes, attribute => attribute != null) : []}
> > ```
> 
> If #1871 is merged this case should be no problem.
> 
> Would be great to test this change in a real-world project.

This links to the wrong change (because differetn repository), this https://github.com/neos/flow-development-collection/pull/1871 makes more sense

**Comment by Sebobo** [comment]
@albe should be put this also into 5.2?

**Comment by albe** [comment]
Did not read through all the conversation, so I'm not sure if this is ready. But if you say it is, I'm happy to pull it into 5.2 now :)

**Comment by hlubek** [comment]
@albe I'd say it should be ready

**Comment by albe** [comment]
@mficzel can you take another look and/or remove your veto?

File: Neos.Fusion/Classes/FusionObjects/ComponentImplementation.php [hunk 2]
```diff
@@ -51,37 +52,32 @@ public function evaluate()
      * @param array $context
      * @return array
      */
-    protected function prepare($context)
+    protected function prepare(array $context): array
```
Comment (albe) [inline Neos.Fusion/Classes/FusionObjects/ComponentImplementation.php:55]: Q: Is this supposed to be extended/inherited and hence the added typehints to be considered breaking?

Comment (mficzel) [reply Neos.Fusion/Classes/FusionObjects/ComponentImplementation.php:55]: This is not @api and no intended extension point. We only seperated this into a special method so we can address this in aspects seperately.

So for me this is ok.

File: Neos.Fusion/Classes/FusionObjects/Helpers/LazyProps.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            } finally {
+                $this->runtime->popContext();
+            }
+        }
+        return $this->valueCache[$path];
+    }
+
+    public function offsetSet($path, $value)
+    {
+        // NOOP
```
Comment (bwaidelich) [inline Neos.Fusion/Classes/FusionObjects/Helpers/LazyProps.php:80]: Shouldn't that throw?

Comment (hlubek) [reply Neos.Fusion/Classes/FusionObjects/Helpers/LazyProps.php:80]: You mean the "set call" should throw an exception? Yes, we could do that.

**Comment by hlubek** [comment]
> Finally tested again and could not construct a fusion problem with this approach any more. All nesting and rendering works as expected and i could not come up with an edge case to break this.

👍 Love to hear that ;)

> 
> However there are two issues when debugging these new proxies that would be a problem for developers:
> 
>     * `{Json.stringify(props)}` >> returns empty (could be fixed by implementing JsonSearializable)

Good idea, that should be easy to implement.

> 
>     * `<Neos.Fusion:Debug>{props}</Neos.Fusion:Debug>` >> runs into errors as the debug cannot be generated
> 
> 
> I could force both by using `{Array.concat(props,[])}` instead of `props` but this is not a solution we can really recommend.
> 
> I have no idea how the second issue can be solved but it should be somehow.
> @hlubek if you have an idea how to solve the debug issues i would really like to get this in.

Maybe we can also use a test for `instanceof JsonSerializable` in the `DebugImplementation` or the Flow debug function itself and use that to extract "public" attributes.

**Comment by mficzel** [comment]
> Maybe we can also use a test for instanceof JsonSerializable in the DebugImplementation or the Flow debug function itself and use that to extract "public" attributes.

Sounds smart, we should try this.

**Comment by mficzel** [comment]
@hlubek i think a have found a more generic approach for the debugging problem https://github.com/neos/flow-development-collection/pull/1980 that would only leave the jsonSerialize open.

**Comment by hlubek** [comment]
@mficzel Thanks for the investigation regarding debugging. I added `JsonSerializable` to `LazyProps` and added a unit test for it. I also confirmed that `Debugger` now works as expected.

Should be good to go now.

**Comment by mficzel** [comment]
@albe @bwaidelich i think this is ready

File: Neos.Fusion/Classes/FusionObjects/Helpers/LazyProps.php (hunk 3 above)
Comment (bwaidelich) [reply Neos.Fusion/Classes/FusionObjects/Helpers/LazyProps.php:80]: I _still_ think trying to set a lazy prop should throw :)

**Comment by albe** [comment]
Thanks everyone! Very nice feature :)

**Comment by hlubek** [comment]
Thanks for reviewing and taking care!

File: Neos.Fusion/Classes/FusionObjects/Helpers/LazyProps.php (hunk 3 above)
Comment (bwaidelich) [reply Neos.Fusion/Classes/FusionObjects/Helpers/LazyProps.php:80]: @albe It's a good idea to resolve all conversations before hitting the merge button. Otherwise open discussions might get lost – like this one.. :)
